packages/block-editor/src/store/reducer.js

```diff
@@ -2131,6 +2140,524 @@ const combinedReducers = combineReducers( {
+ * @param {Function} callback Function to execute for each block encountered during traversal.
+ *                            The callback receives the current block as its argument.
+ */
+function traverseBlockTree( state, clientId, callback ) {
+	const parentTree = getBlockTreeBlock( state, clientId );
+	if ( ! parentTree ) {
+		return;
+	}
+
+	callback( parentTree );
+
+	if ( ! parentTree?.innerBlocks?.length ) {
+		return;
+	}
+
+	for ( const block of parentTree?.innerBlocks ) {
+		traverseBlockTree( state, block.clientId, callback );
+	}
+}
+
+/**
+ * Checks if a block has a parent in a list of client IDs, and if so returns the client ID of the parent.
+ *
+ * @param {Object} state     The current state object.
+ * @param {string} clientId  The client ID of the block to search the parents of.
+ * @param {Array}  clientIds The client IDs of the blocks to check.
+ *
+ * @return {string|undefined} The client ID of the parent block if found, undefined otherwise.
+ */
+function findParentInClientIdsList( state, clientId, clientIds ) {
+	let parent = state.blocks.parents.get( clientId );
+	while ( parent ) {
+		if ( clientIds.includes( parent ) ) {
+			return parent;
+		}
+		parent = state.blocks.parents.get( parent );
+	}
+}
+
+/**
+ * Checks if a block has any bindings in its metadata attributes.
+ *
+ * @param {Object} block The block object to check for bindings.
+ * @return {boolean}    True if the block has bindings, false otherwise.
+ */
+function hasBindings( block ) {
+	return (
+		block?.attributes?.metadata?.bindings &&
+		Object.keys( block?.attributes?.metadata?.bindings ).length
+	);
+}
+
+/**
+ * Determines the editing mode for a given block based on various factors.
+ *
+ * This function evaluates the editing mode of a block considering the following:
+ * - Whether the editor is in zoomed out or navigation mode.
+ * - The block's relationship to sections and synced patterns.
+ * - The presence of bindings in the block's attributes.
+ *
+ * @param {Object}  state                          The current state object.
+ * @param {string}  clientId                       The client ID of the block to evaluate.
+ * @param {Object}  options                        Additional options for determining the editing mode.
+ * @param {string}  options.sectionRootClientId    The client ID of the section root block.
+ * @param {Array}   options.sectionClientIds       An array of client IDs for section blocks.
+ * @param {Array}   options.syncedPatternClientIds An array of client IDs for synced pattern blocks.
+ * @param {boolean} options.isZoomedOut            Whether the editor is in zoomed out mode.
+ * @param {boolean} options.isNavMode              Whether the editor is in navigation mode.
+ *
+ * @return {string|undefined} The determined editing mode for the block: 'contentOnly', 'disabled', or undefined.
+ */
+function getDerivedBlockEditingModeForBlock(
+	state,
+	clientId,
+	{
```

Comment:
What's the difference between this argument and all the content that is already in `state`?

Reply:
It's mostly stuff that's already been retrieved or calculated by the calling function, so it's an optimization to pass it in rather than determine it again for every block.

In an earlier iteration this function was called from a couple of different places, but now it's only called from `getDerivedBlockEditingModesForTree` so I could inline it there.

It'd make a small dent in the line of code changed in this PR 😆

Reply:
They seem to need to be derrived in various ways since they're not preset in this state?

Reply:
If the optimization is helpful, I don't mind it, but maybe at least at a comment.

Reply:
I've inlined this code now so no comment needed.